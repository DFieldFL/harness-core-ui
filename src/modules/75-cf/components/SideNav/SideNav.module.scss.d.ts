/* eslint-disable */
// this is an auto-generated file
declare const styles: {
  readonly projectSetup: string
}
export default styles
