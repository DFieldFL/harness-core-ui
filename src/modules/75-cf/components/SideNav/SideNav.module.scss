.onboarding {
  --layout-spacing: var(--spacing-xlarge);
  position: relative;

  &:before {
    position: absolute;
    content: '';
    height: 1px;
    width: 180px;
    top: -13px;
    left: -15px;
    background-color: rgba(255, 255, 255, 0.1);
  }
}
