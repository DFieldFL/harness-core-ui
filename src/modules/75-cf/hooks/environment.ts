import type { SelectOption } from '@wings-software/uicore'
import { useStrings } from 'framework/exports'
import type { StringKeys } from 'framework/strings/StringsContext'
import { GetEnvironmentListForProjectQueryParams, useGetEnvironmentListForProject } from 'services/cd-ng'

export const useEnvironments = (queryParams: GetEnvironmentListForProjectQueryParams) => {
  const { data: environments, loading, error, refetch } = useGetEnvironmentListForProject({
    queryParams
  })

  const convertToSelectOptions: SelectOption[] =
    environments?.data?.content?.map<SelectOption>(elem => ({
      label: elem.name as string,
      value: elem.identifier as string
    })) || []

  return {
    data: convertToSelectOptions,
    loading,
    error,
    refetch
  }
}

/**
 * getEnvString prepends "cf.environments" to given argument
 * Provides base getString in case of needing it
 */
export const useEnvStrings = () => {
  const { getString } = useStrings()
  return {
    getEnvString: (key: string, vars?: Record<string, any>) =>
      getString(`cf.environments.${key}` as StringKeys /* TODO: fix this by using a map */, vars),
    getString
  }
}
