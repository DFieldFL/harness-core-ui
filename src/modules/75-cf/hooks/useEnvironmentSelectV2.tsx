import React, { useEffect, useState } from 'react'
import { useParams } from 'react-router-dom'
import { Select, SelectOption, SelectProps } from '@wings-software/uicore'
import { EnvironmentResponseDTO, useGetEnvironmentListForProject } from 'services/cd-ng'

export interface UseEnvironmentSelectV2Params {
  selectedEnvironmentIdentifier?: string
  onChange: (opt: SelectOption, environment: EnvironmentResponseDTO, userEvent: boolean) => void
}

export const useEnvironmentSelectV2 = (params: UseEnvironmentSelectV2Params) => {
  const { onChange, selectedEnvironmentIdentifier } = params
  const { projectIdentifier, orgIdentifier, accountId } = useParams<Record<string, string>>()
  const { data, loading, error, refetch } = useGetEnvironmentListForProject({
    queryParams: { accountId, orgIdentifier, projectIdentifier }
  })
  const [selectedEnvironment, setSelectedEnvironment] = useState<SelectOption>()
  const selectOptions: SelectOption[] =
    data?.data?.content?.map<SelectOption>(elem => ({
      label: elem.name as string,
      value: elem.identifier as string
    })) || []

  useEffect(() => {
    if (data?.data?.content?.length) {
      if (selectedEnvironmentIdentifier) {
        const found = data?.data?.content?.find(env => env.identifier === selectedEnvironmentIdentifier)

        if (found) {
          const newValue = {
            label: found.name as string,
            value: found.identifier as string
          }
          setSelectedEnvironment(newValue)
          onChange(newValue, found, false)
          return
        }
      }

      setSelectedEnvironment(selectOptions[0])
      onChange(selectOptions[0], data?.data?.content?.[0], false)
    }
  }, [data?.data?.content?.length, data?.data?.content?.find, selectedEnvironmentIdentifier]) // eslint-disable-line

  return {
    EnvironmentSelect: function EnvironmentSelect(props: Partial<SelectProps>) {
      return (
        <Select
          value={selectedEnvironment}
          items={selectOptions}
          onChange={opt => {
            if (selectedEnvironment?.value !== opt.value) {
              setSelectedEnvironment(opt)
              onChange(
                opt,
                data?.data?.content?.find(env => env.identifier === opt.value) as EnvironmentResponseDTO,
                true
              )
            }
          }}
          {...props}
        />
      )
    },
    loading,
    error,
    refetch,
    environments: data?.data?.content
  }
}
