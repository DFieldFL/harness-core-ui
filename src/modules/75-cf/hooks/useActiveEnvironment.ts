import { useQueryParams } from '@common/hooks'

const useActiveEnvironment = (): {
  activeEnvironment: string
  withActiveEnvironment: (url: string, envOverride?: string) => string
} => {
  const activeEnvironment = useQueryParams<Record<string, string>>()?.activeEnvironment || ''

  const withActiveEnvironment = (url: string, envOverride?: string): string => {
    const env = envOverride ?? activeEnvironment
    if (!env || url.includes(`activeEnvironment=${env}`)) return url

    if (url.includes('activeEnvironment')) return url.replace(/activeEnvironment=[^&]+/gi, `activeEnvironment=${env}`)

    return `${url}${url.includes('?') ? '&' : '?'}activeEnvironment=${env}`
  }

  return {
    activeEnvironment: activeEnvironment !== 'undefined' ? activeEnvironment : '',
    withActiveEnvironment
  }
}

export default useActiveEnvironment
