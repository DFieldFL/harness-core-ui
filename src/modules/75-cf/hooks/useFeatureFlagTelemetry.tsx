import { useTelemetry } from '@common/hooks/useTelemetry'
import { ModuleName } from 'framework/types/ModuleName'

interface FeatureFlagTelementary {
  visitedPage: () => void
  createFeatureFlagStart: () => void
  createFeatureFlagCompleted: () => void
}

const EVENT_TYPE = {
  VISITED_LANDING_PAGE: 'Visited feature flag landing page',
  CREATE_FEATURE_FLAG_START: 'Started creating feature flag',
  CREATE_FEATURE_FLAG_COMPLETE: 'Successfully created feature flag'
}

export const useFeatureFlagTelemetry = (): FeatureFlagTelementary => {
  const { trackEvent } = useTelemetry()

  const visitedPage = (): void => trackEvent(EVENT_TYPE.VISITED_LANDING_PAGE, { module: ModuleName.CF })
  const createFeatureFlagStart = (): void => trackEvent(EVENT_TYPE.CREATE_FEATURE_FLAG_START, { module: ModuleName.CF })
  const createFeatureFlagCompleted = (): void =>
    trackEvent(EVENT_TYPE.CREATE_FEATURE_FLAG_COMPLETE, { module: ModuleName.CF })

  return { visitedPage, createFeatureFlagStart, createFeatureFlagCompleted }
}
