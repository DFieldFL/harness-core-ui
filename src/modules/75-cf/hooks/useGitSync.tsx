import { useParams } from 'react-router-dom'
import * as yup from 'yup'
import type { ObjectSchema } from 'yup'
import type { ModulePathParams, ProjectPathProps } from '@common/interfaces/RouteInterfaces'
import { GitRepo, useGetGitRepo, usePatchGitRepo } from 'services/cf'
import { useFeatureFlag } from '@common/hooks/useFeatureFlag'
import { FeatureFlag } from '@common/featureFlags'
import { useStrings } from 'framework/strings'

export interface GitDetails {
  branch: string
  filePath: string
  repoIdentifier: string
  rootFolder: string
  commitMsg: string
}
export interface GitSyncFormValues {
  gitDetails: GitDetails
  autoCommit: boolean
}
interface GitSyncFormMeta {
  gitSyncInitialValues: GitSyncFormValues
  gitSyncValidationSchema: ObjectSchema<Record<string, unknown> | undefined>
}

export interface UseGitSync {
  gitRepoDetails: GitRepo | undefined
  isAutoCommitEnabled: boolean
  isGitSyncEnabled: boolean | undefined
  gitSyncLoading: boolean
  handleAutoCommit: (newAutoCommitValue: boolean) => Promise<void>
  getGitSyncFormMeta: (autoCommitMessage?: string) => GitSyncFormMeta
}

export const useGitSync = (): UseGitSync => {
  const { projectIdentifier, accountId, orgIdentifier } = useParams<ProjectPathProps & ModulePathParams>()
  const { getString } = useStrings()

  const getGitRepo = useGetGitRepo({
    identifier: projectIdentifier,
    queryParams: {
      accountIdentifier: accountId,
      org: orgIdentifier
    }
  })

  const patchGitRepo = usePatchGitRepo({
    identifier: projectIdentifier,
    queryParams: {
      accountIdentifier: accountId,
      org: orgIdentifier
    }
  })

  const FF_GITSYNC = useFeatureFlag(FeatureFlag.FF_GITSYNC)

  const isGitSyncEnabled = FF_GITSYNC && getGitRepo?.data?.repoSet

  const isAutoCommitEnabled = (isGitSyncEnabled && getGitRepo?.data?.repoDetails?.autoCommit) || false

  const getGitSyncFormMeta = (autoCommitMessage?: string): GitSyncFormMeta => ({
    gitSyncInitialValues: {
      gitDetails: {
        branch: getGitRepo?.data?.repoDetails?.branch || '',
        filePath: getGitRepo?.data?.repoDetails?.filePath || '',
        repoIdentifier: getGitRepo?.data?.repoDetails?.repoIdentifier || '',
        rootFolder: getGitRepo?.data?.repoDetails?.rootFolder || '',
        commitMsg: isAutoCommitEnabled && autoCommitMessage ? `[AUTO-COMMIT] : ${autoCommitMessage}` : ''
      },
      autoCommit: isAutoCommitEnabled
    },
    gitSyncValidationSchema: yup.object().shape({
      commitMsg: isGitSyncEnabled ? yup.string().required(getString('cf.creationModal.commitMsg')) : yup.string()
    })
  })

  const handleAutoCommit = async (newAutoCommitValue: boolean): Promise<void> => {
    if (newAutoCommitValue && isAutoCommitEnabled != newAutoCommitValue) {
      const instruction = {
        instructions: [
          {
            kind: 'setAutoCommit',
            parameters: {
              autoCommit: newAutoCommitValue
            }
          }
        ]
      }

      await patchGitRepo.mutate(instruction)
      await getGitRepo.refetch()
    }
  }

  return {
    gitRepoDetails: getGitRepo?.data?.repoDetails,
    isAutoCommitEnabled,
    isGitSyncEnabled,
    gitSyncLoading: getGitRepo?.loading || patchGitRepo.loading,
    handleAutoCommit,
    getGitSyncFormMeta
  }
}
