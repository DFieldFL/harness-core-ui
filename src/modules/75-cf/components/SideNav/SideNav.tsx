import React from 'react'
import { useParams, useHistory } from 'react-router-dom'
import { Layout } from '@wings-software/uicore'
import routes from '@common/RouteDefinitions'
import { ProjectSelector } from '@common/navigation/ProjectSelector/ProjectSelector'
import type { PipelinePathProps } from '@common/interfaces/RouteInterfaces'
import { SidebarLink } from '@common/navigation/SideNav/SideNav'
import { ModuleName } from 'framework/types/ModuleName'
import { useStrings } from 'framework/strings'
import { useAppStore } from 'framework/AppStore/AppStoreContext'
import useActiveEnvironment from '@cf/hooks/useActiveEnvironment'
import css from './SideNav.module.scss'

export default function CFSideNav(): React.ReactElement {
  const { getString } = useStrings()
  const params = useParams<PipelinePathProps>()
  const { accountId, projectIdentifier, orgIdentifier } = params
  const history = useHistory()
  const module = 'cf'
  const { updateAppStore } = useAppStore()
  const isDev = localStorage.ENABLED_FF_EXPERIMENTS
  const { withActiveEnvironment } = useActiveEnvironment()

  return (
    <Layout.Vertical spacing="small">
      <ProjectSelector
        moduleFilter={ModuleName.CF}
        onSelect={data => {
          updateAppStore({ selectedProject: data })
          history.push(
            routes.toCFFeatureFlags({
              projectIdentifier: data.identifier,
              orgIdentifier: data.orgIdentifier || '',
              accountId
            })
          )
        }}
      />
      {projectIdentifier && orgIdentifier ? (
        <React.Fragment>
          <SidebarLink
            label={getString('featureFlagsText')}
            to={withActiveEnvironment(routes.toCFFeatureFlags(params))}
          />
          <SidebarLink
            label={getString('cf.shared.targetManagement')}
            to={withActiveEnvironment(routes.toCFTargetManagement(params))}
          />
          <SidebarLink label={getString('environments')} to={withActiveEnvironment(routes.toCFEnvironments(params))} />
          {isDev && (
            <>
              <SidebarLink
                label={getString('executionsText')}
                to={withActiveEnvironment(routes.toDeployments({ ...params, module }))}
              />
              <SidebarLink
                label={getString('pipelines')}
                to={withActiveEnvironment(routes.toPipelines({ ...params, module }))}
              />
            </>
          )}
          {isDev && (
            <SidebarLink
              className={css.onboarding}
              label={getString('cf.shared.getStarted')}
              to={withActiveEnvironment(routes.toCFOnboarding(params))}
            />
          )}
        </React.Fragment>
      ) : null}
    </Layout.Vertical>
  )
}
