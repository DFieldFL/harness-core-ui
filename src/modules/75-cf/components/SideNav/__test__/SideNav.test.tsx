import React from 'react'
import { fireEvent, render, screen } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import { useFeatureFlagTelemetry } from '@cf/hooks/useFeatureFlagTelemetry'
import * as hooks from '@common/hooks/useFeatureFlag'
import SideNav from '../SideNav'

jest.mock('@cf/hooks/useFeatureFlagTelemetry', () => ({
  useFeatureFlagTelemetry: jest.fn(() => ({
    visitedPage: jest.fn(),
    createFeatureFlagStart: jest.fn(),
    createFeatureFlagCompleted: jest.fn()
  }))
}))

jest.mock('@common/hooks/useQueryParams', () => ({
  useQueryParams: () => jest.fn()
}))

jest.mock('@common/hooks/useTelemetry', () => ({
  useTelemetry: () => ({ identifyUser: jest.fn(), trackEvent: jest.fn() })
}))

describe('Sidenav', () => {
  beforeEach(() => localStorage.clear())

  const Subject: React.FC<{ path?: string }> = ({
    path = '/account/:accountId/cf/dashboard/orgs/:orgIdentifier/projects/:projectIdentifier'
  }) => (
    <TestWrapper path={path} pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy' }}>
      <SideNav />
    </TestWrapper>
  )

  test('it should render', () => {
    const { container } = render(<Subject />)
    expect(container).toMatchSnapshot()
  })

  test('it should display the pipelines link only when pipelines are enabled', async () => {
    const { rerender } = render(<Subject />)
    expect(screen.queryByText('pipelines')).not.toBeInTheDocument()

    localStorage.setItem('FF_PIPELINES', 'true')
    rerender(<Subject />)
    expect(screen.getByText('pipelines')).toBeInTheDocument()
  })

  test('it should hide the Git Experience links when FF_GITSYNC is FALSE', async () => {
    jest.spyOn(hooks, 'useFeatureFlags').mockImplementation(() => ({ FF_GITSYNC: false }))

    render(
      <Subject path="/account/:accountId/cf/orgs/:orgIdentifier/projects/:projectIdentifier/setup/access-control/users" />
    )

    expect(screen.queryByText('connectorsLabel')).not.toBeInTheDocument()
    expect(screen.queryByText('common.secrets')).not.toBeInTheDocument()
  })

  test('it should show the Git Experience links when FF_GITSYNC is TRUE', async () => {
    jest.spyOn(hooks, 'useFeatureFlags').mockImplementation(() => ({ FF_GITSYNC: true }))

    render(
      <Subject path="/account/:accountId/cf/orgs/:orgIdentifier/projects/:projectIdentifier/setup/access-control/users" />
    )

    expect(screen.getByText('connectorsLabel')).toBeInTheDocument()
    expect(screen.queryByText('common.secrets')).toBeInTheDocument()
  })

  test('it should fire telementary event when Feature Flags menu item clicked', () => {
    render(<Subject />)

    const featureFlagLink = screen.getByText('featureFlagsText')
    expect(featureFlagLink).toBeInTheDocument()

    fireEvent.click(featureFlagLink)

    expect(useFeatureFlagTelemetry).toHaveBeenCalled()
  })
})
