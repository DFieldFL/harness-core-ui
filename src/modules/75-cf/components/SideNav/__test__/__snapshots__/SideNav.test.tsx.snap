// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidenav render 1`] = `
<div>
  <div
    class="Layout--vertical Layout--layout-spacing-small StyledProps--main"
  >
    <div
      class="Layout--vertical StyledProps--main StyledProps--padding-large"
    >
      <p
        class="StyledProps--font StyledProps--main StyledProps--margin-bottom-xsmall StyledProps--font-size-small StyledProps--color StyledProps--color-grey500"
      >
        projectLabel
      </p>
      <div
        class="projectSelector selectProjectDisplay"
      >
        <div
          class="bp3-popover-wrapper bp3-fill"
        >
          <div
            class="bp3-popover-target"
          >
            <button
              class="bp3-button bp3-minimal popoverTarget selectButton"
              data-testid="project-select-button"
              type="button"
            >
              <span
                class="bp3-button-text"
              >
                <p
                  class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey400 StyledProps--font-size-normal StyledProps--padding-xsmall"
                >
                  selectProject
                </p>
              </span>
              <span
                class="bp3-icon bp3-icon-chevron-right"
                icon="chevron-right"
              >
                <svg
                  data-icon="chevron-right"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <desc>
                    chevron-right
                  </desc>
                  <path
                    d="M10.71 7.29l-4-4a1.003 1.003 0 00-1.42 1.42L8.59 8 5.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l4-4c.18-.18.29-.43.29-.71 0-.28-.11-.53-.29-.71z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pickProjectHelp"
      style="background-image: url(test-file-stub);"
    >
      pickProject
    </div>
    <a
      class="link"
      href="/account/dummy/cf/orgs/dummy/projects/dummy/feature-flags"
    >
      <p
        class="StyledProps--font StyledProps--main text"
      >
        featureFlagsText
      </p>
    </a>
    <a
      class="link"
      href="/account/dummy/cf/orgs/dummy/projects/dummy/target-management"
    >
      <p
        class="StyledProps--font StyledProps--main text"
      >
        cf.shared.targetManagement
      </p>
    </a>
    <a
      class="link"
      href="/account/dummy/cf/orgs/dummy/projects/dummy/environments"
    >
      <p
        class="StyledProps--font StyledProps--main text"
      >
        environments
      </p>
    </a>
    <a
      class="link onboarding"
      href="/account/dummy/cf/orgs/dummy/projects/dummy/onboarding"
    >
      <p
        class="StyledProps--font StyledProps--main text"
      >
        cf.shared.getStarted
      </p>
    </a>
    <div
      class="Layout--vertical StyledProps--main main"
    >
      <p
        class="StyledProps--font StyledProps--main text StyledProps--font-xsmall StyledProps--flex StyledProps--flex-alignItems-flex-start StyledProps--flex-justifyContent-space-between"
        style="display: flex; align-items: center;"
      >
        common.projectSetup
        <span
          class="bp3-icon bp3-icon-chevron-up StyledProps--main StyledProps--padding-left-xsmall"
          icon="chevron-up"
        >
          <svg
            data-icon="chevron-up"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <desc>
              chevron-up
            </desc>
            <path
              d="M12.71 9.29l-4-4C8.53 5.11 8.28 5 8 5s-.53.11-.71.29l-4 4a1.003 1.003 0 001.42 1.42L8 7.41l3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </p>
    </div>
  </div>
</div>
`;
